src/snarky-typescript/src/index.ts

```diff
@@ -0,0 +1,517 @@
+  div(x: Var | F, y: Var | F): Var;
+
+  scale(x: Var | F, scalar: FieldConvertible): Var;
+  scale(x: Var | F, scalar: F): Var;
+
+  equals(x: FieldConvertible, y: FieldConvertible): Var;
+  equals(x: FieldConvertible, y: Var | F): Var;
+  equals(x: Var | F, y: FieldConvertible): Var;
+  equals(x: Var | F, y: Var | F): Var;
+
+  assertBoolean(x: FieldConvertible): BoolVar<F, Circuit, Var>;
+  assertBoolean(x: Var | F): BoolVar<F, Circuit, Var>;
+  assertBoolean(x: Var): BoolVar<F, Circuit, Var>;
+
+  assertEquals(x: FieldConvertible, y: FieldConvertible): void;
+  assertEquals(x: FieldConvertible, y: Var | F): void;
+  assertEquals(x: Var | F, y: FieldConvertible): void;
+  assertEquals(x: Var | F, y: Var | F): void;
+
+  toBits(x: FieldConvertible): BoolVar<F, Circuit, Var>[];
+  toBits(x: Var | F): BoolVar<F, Circuit, Var>[];
+
+  ofBits(x: (boolean | BoolVar<F, Circuit, Var>)[]): Var;
+
+  toFieldElements(x: FieldConvertible): [Var];
+  toFieldElements(x: Var | F): [Var];
+
+  ofFieldElements(x: (FieldConvertible | F | Var)[]): Var;
+
+  store(x: Var, val: FieldConvertible): void;
+  store(x: Var, val: F): void;
+  store(x: Var, f: () => F): void;
+  store(x: Var, f: () => FieldConvertible): void;
+
+  read(x: FieldConvertible): F;
+  read(x: F): F;
+  read(x: Var): F;
+
+  new (): Var;
+  new (options: {isPublic: boolean}): Var;
+
+  new (x: FieldConvertible): Var;
+  new (x: F): Var;
+  new (x: Var): Var;
+  new (f: () => FieldConvertible | F | Var): Var;
+
+  new (x: FieldConvertible, options: {isPublic: boolean}): Var;
+  new (x: F, options: {isPublic: boolean}): Var;
+  new (x: Var, options: {isPublic: boolean}): Var;
+  new (f: () => F, options: {isPublic: boolean}): Var;
+  new (f: () => FieldConvertible, options: {isPublic: boolean}): Var;
+
+  new (x: FieldConvertible, isPublic: boolean): Var;
+  new (x: F, isPublic: boolean): Var;
+  new (x: Var, isPublic: boolean): Var;
+  new (f: () => F, isPublic: boolean): Var;
+  new (f: () => FieldConvertible, isPublic: boolean): Var;
+}
+
+/* TODO: Fill this in with the default implementation */
+abstract class BoolVar<F, Circuit, V extends CircuitVar<F, Circuit>> {
+  abstract not(this: this): this;
+  abstract and(this: this, y: boolean | this): this;
+  abstract or(this: this, y: boolean | this): this;
+
+  abstract equals(this: this, y: boolean | this): this;
+  abstract assertEquals(this: this, y: boolean | this): void;
+  abstract assertTrue(this: this): void;
+  abstract assertFalse(this: this): void;
+
+  abstract if_<Var extends AsFieldElements<V, Var>>(
+    this: this,
+    then_: Var,
+    else_: Var
+  ): Var;
+  abstract if_<Var extends AsFieldElements<V, Var>>(
+    this: this,
+    args: {
+      then_: Var;
+      else_: Var;
+    }
+  ): Var;
+
+  abstract toField(this: this): V;
+  abstract toFieldElements(this: this): [V];
+  abstract ofFieldElements(x: V[]): this;
+
+  abstract read(this: this): boolean;
+}
+
+interface BoolVarCtor<
```

Comment:
(like above), doesn't this interface `extend AsFieldElements`? 

This pattern continues, but I will refrain from commenting on each occurrence.